### PR TITLE
updating to build with newest rust

### PR DESCRIPTION
**Pull request by waynenilsen**
I'm new to rust but not scientific computing and I wanted to use this library. It wouldn't build so I forked it and went through many of the build errors and warnings trying to shut them up. It still does not compile but it is much further along than when I started. The first commit should be good but the second one had a bunch of find and replace stuff (in a directory) which usually gets messy.

### CODE REVIEW

File: src/alg/sort/insertion_sort.rs [hunk 1]
```diff
@@ -8,7 +8,7 @@ use std::ptr;
 pub fn insertion_sort_slice<T : PartialOrd>(data : &mut [T]){
     let n  = data.len();
     for j in range(1, n){
-        // we insert data[j] into the sorted sequence 
+        // we insert data[j] isizeo the sorted sequence
```
Comment (waynenilsen) [inline src/alg/sort/insertion_sort.rs]: some of the docs got mangled, obviously those wont get merged

**Comment by shailesh1729** [comment]
Hi waynenilsen,

I had been working on this library in the last quarter of the 2015. While I left for a couple of weeks, Rust changed a lot and it introduced huge amount of  build errors. Since then I have not had the courage to sit down and fix it all. Now that you are the first person interested in looking at  it,  I think I will sit down and resolve pending issues. Moreover, we can discuss more about the library. Feel free to contact me over email: shailesh at indigits dot com.

**Comment by waynenilsen** [comment]
Shailesh, I hope that the changes I made make sense and you can use them to help bring this library up to the newest Rust version. The `cargo build` command is an interesting thing, in c++ i'm used to fixing all the things that the compiler is complaining about, then it just compiles. In Rust, you think you're almost done, fix the last thing the compiler is complaining about, then it comes up with 100 more things to complain about haha. Anyway, as I am new to Rust, I wasn't sure how to fix the latest thing it was annoyed about, something like `Self` vs expecting `&Self` or visa versa.

**Comment by shailesh1729** [comment]
waynenilsen,  The merge is going to be complicated. I had some changes sitting at my end too. So I  am going to do the merge manually going file by file. At this moment, the library has started building but I still have to finish the unit tests part.  Meanwhile, I am struggling with a particular language issue. It's about making the operator overloading work nicely for reference types. See my post on Reddit http://www.reddit.com/r/rust/comments/2wq7qm/converting_from_t_to_t_at_function_call_site_for/ . If you have any specific suggestion, do let me know.

**Comment by waynenilsen** [comment]
I was thinking something like why not just return a reference to the value such as 

```rust
use std::ops::Add;

#[derive(Debug)]
struct A {
  x: i32,
  y : i32
}


impl<'a, 'b> Add<&'b A> for &'a A{
    type Output = &'b A;
    fn add(self, rhs: &'b A) -> &'b A {
        & A{x : self.x + rhs.x, y : self.y + rhs.y}
    }
}

fn main() {
    let a = A {x : 1, y : 2};
    let b = A {x : 1, y : 2};
    let c = A {x : 1, y : 2};
    let d = *(&a + &b + &c);
    println!("{:?}", c);
}
```

but I can't get it to compile, see [http://is.gd/B7OA8x](http://is.gd/B7OA8x) I have to say I'm new to rust and don't really understand this whole borrowing thing, I get 

`borrowed value does not live long enough`

**Comment by shailesh1729** [comment]
The build is up now and most unit tests are running fine. Some more cleanup is still pending. Complex matrices are not working at the moment.

**Comment by shailesh1729** [comment]
Let me explain why your idea cannot work.  The result of sum is created on stack as a local variable. Reference to a local variable cannot be returned. When you return by value, then Rust implements MOVE semantics. The contents of the variable on the stack are moved into the variable receiving the result.

**Comment by shailesh1729** [comment]
Latest scirust build and test run results can be seen at https://travis-ci.org/indigits/scirust. It looks like Rust compiler is still cleaning up lots of things in usage. As a result builds are breaking regularly. So can't guarantee if this code will remain compilable for long. I don't change my personal Rust daily. Only when the builds start failing, I consider changing.

**Comment by waynenilsen** [comment]
Great work, thanks for explaining the move semantics I totally understand now why you cant do what I suggested. Did you end up using any of the changes from my fork?

**Comment by shailesh1729** [comment]
For the overriding of add for expressions, following suggestion seems to be  better. At least now only one ampersand is needed for each variable in a long expression. Ampersends are not needed for temporaries. 

    use std::ops::Add;
    
    #[derive(Debug)]
    struct A {
      x: i32,
      y : i32
    }
    
    
    impl<'a, 'b> Add<&'b A> for &'a A{
        type Output = A;
        fn add(self, rhs: &'b A) -> A{
            A{x : self.x + rhs.x, y : self.y + rhs.y}
        }
    }
    
    impl<'b> Add<&'b A> for A{
        type Output = A;
        fn add(self, rhs: &'b A) -> A{
            &self + rhs
        }
    }
    
    
    
    fn main() {
        let a = A {x : 1, y : 2};
        let b = A {x : 1, y : 2};
        let c = &a + &b + &b + &b + &b;
        println!("{:?}", c);
    }

**Comment by shailesh1729** [comment]
Items from the pull request.

- The int/uint to isize/usize changes have been merged pretty much with few exceptions (noted below).
- did migrate to rand library from crates.io 
- examined each find/replace carefully to ensure that documentation etc. aren't affected. 
- language changes taken care of.
- Iterators have been properly worked out by defining Item associated type carefully. 


Some differences
- removed MatrixUInt altogether. Now, I don't think it's useful. Better to work with matrices with fixed size data types. Similarly removed similar type definitions in other modules. 
- Wherever uint/int were used for rows, columns, locations, offsets, etc. they have been replaced with usize and isize.But if they were used for underlying data type for a Vec or a Matrix, they being changed with some fixed integer type. 


On the merge side, I went slightly differently. Rust compiler has this tendency that it doesn't throw all errors and warnings in one go.  Sometimes only few errors are thrown, then once u clean them u see a whole lot of new errors popping up. I haven't yet figured out it's exact behavior. But usually if there is some import problem, then it stops there. Once all import problems are cleared then it starts throwing other compilation problems.  These new warnings due to feature depredations (like int being replaced by isize etc.) although interesting but were coming in the way of the real work. Fortunately they provided some crate level attributes like: #![feature(int_uint)], using which these warnings could be ignored. I first ignored those warnings, and focused on compiling everything correctly. A number of other things had to be fixed since Trait behavior has changed, several APIs like Vector etc. have changed. all good improvements in the language but had to spend time figuring out the changes and getting them right in the code.

**Pull request closed without merging**